12-offer-encoding.md

```diff
@@ -0,0 +1,799 @@
+    - the bitcoin chain is implied as the first and only entry.
+  - if it has bolt11 features:
+    - MUST set `features` to the bitmap of features.
+  - if the expiry for accepting payment is not 7200 seconds after `created_at`:
+    - MUST set `relative_expiry` `seconds_from_creation` to the number of
+      seconds after `created_at` that payment of this invoice should not be attempted.
+  - if the `min_final_cltv_expiry` for the last HTLC in the route is not 18:
+    - MUST set `min_final_cltv_expiry`.
+  - if it accepts onchain payments:
+    - MAY specify `fallbacks`
+    - MUST specify `fallbacks` in order of most-preferred to least-preferred
+      if it has a preference.
+    - for the bitcoin chain, it MUST set each `fallback_address` with
+      `version` as a valid witness version and `address` as a valid witness
+      program
+  - if it is connected only by private channels:
+    - if no node connecting a private channel supports blinded payments:
+      - MUST include `desthint` containing the private channel information.
+    - otherwise:
+      - MUST include `blinded_path` containing one or more paths to the node.
+  - otherwise:
+    - MAY include `blinded_path`.
+  - if it includes `blinded_path`:
+    - MUST specify `path` in order of most-preferred to least-preferred if
+      it has a preference.
+    - MUST include `blindedpay` with exactly one `payinfo` for
+      each `blinded_path` in `paths`, in order.
+    - if it includes `blinded_capacities`:
+      - MUST include exactly one `incoming_msat` (in millisatoshis) per `blinded_path`, reflecting the expected maximum amount that can be sent through the path.
+    - SHOULD ignore any payment which does not use one of the paths.
+  - otherwise:
+    - MUST NOT include `blinded_payinfo`.
+  - MUST specify `amount`.`msat` in multiples of the minimum lightning-payable unit
+    (e.g. milli-satoshis for bitcoin) for `chain` (or for bitcoin, if there is no `chain`).
+  - if responding to an `invoice_request`:
+    - if for the same `offer_id` and `payer_key` as a previous `invoice_request`:
+      - MAY simply reuse the previous invoice.
+    - otherwise:
+      - MUST NOT reuse a previous invoice.
+    - MUST set `node_id` the same as the offer.
+    - MUST set (or not set) `quantity` exactly as the invoice_request did.
+    - MUST set `payer_key` exactly as the invoice_request did.
+    - MUST set (or not set) `payer_info` exactly as the invoice_request did.
+    - MUST set (or not set) `payer_note` exactly as the invoice_request did,
+      or MUST not set it.
+    - MUST set (or not set) `replace_invoice` exactly as the invoice_request did.
+    - MUST begin `description` with the `description` from the offer.
+    - MAY append additional information to `description` (e.g. " +shipping").
+    - if it does not set `amount` to the *base invoice amount* calculated from the invoice_request:
+       - MUST append the reason to `description` (e.g. " 5% bulk discount").
```

Comment:
I don't think we should change the description or the amount here.
The offer should lay out all the terms of the payment and the invoice request should be seen as an agreement with the terms of the offer and a willingness to pay immediately. Everything that comes after (the invoice and the payment itself) should be technicalities that can stay hidden from the user.
If you think it's desirable to add extra costs for shipping or bulk discount, these should be part of the offer.

Reply:
In general, you cannot do this.  For example, you don't know shipping costs until after presented (future extension) with the address in the invoice_request.

Of course, it's *unusual*, and hence the requirement to annotate the description.

Reply:
If it's for a future extension, then it should be added in the future extension, not now.

It also means that the amount in the invoice may be completely different from the amount in the request, it goes against what you wrote in the rationale of the invoice request:

> Users can give a tip (or obscure the amount sent) by specifying an amount in their invoice request, even though the offer specifies an amount.

I've rounded the amount to increase my privacy and now the invoice asks me to pay an amount that would leak my shipping address!

Reply:
OK, let's bar differences for now.  It's probable people will start doing this later (despite any wording we add to disallow it here), and they'll break things, and we'll have to re-add it...